priv/blogs/2022/07-29-observing-elixir-with-lightstep.md

```diff
@@ -0,0 +1,222 @@
+
+To start using these libraries, most simply require calling a `setup/1` function in our Application supervision tree.
+
+```elixir
+# my_app/lib/my_app/application.ex
+defmodule MyApp.Application do
+  use Application
+ 
+  def start(_type, _args) do
+    OpentelemetryPhoenix.setup()
+    OpentelemetryLiveView.setup()
+    OpentelemetryEcto.setup([:my_app, :repo])
+
+    # Only trace jobs to minimize noise
+    OpentelemetryOban.setup(trace: [:jobs])
+  
+    ...
+  end
+end
+```
+
+### Instrumenting your application
+
+With distributed tracing, it is recommended to start by tracing the [boundaries of your application](https://vi.to/hubs/o11yfest/videos/3149?v=%2Fvideos%2F3149). As we saw above, we can get spans into many of the entrypoints of our application just by using out-of-the-box libraries. However, there may be internal operations that may need spans, or maybe you want to add additional information to spans that are already being produced, such as adding the `user_id` to every span in your [plug](https://hexdocs.pm/plug/readme.html) pipeline.
+
+### Configuration
+
+Before we start sending data, we need to add a few configurations to our app
+
+1. Run the `opentelemetry` application as temporary
+
+Although observability is important, if it fails it shouldn't take your app down with it! To ensure this is the case, we need to make sure `openteletry` runs in temporary mode
+
+```elixir
+# mix.exs
+
+def project do
+  [
+    ...,
+    releases: [
+      my_app: [
+          applications: [my_app: :permanent, opentelemetry: :temporary]
+      ]
+    ]
+  ]
+end
+```
+
+2. Set your runtime specific attributes for your production environment
+
+My application currently runs in [fly.io](https://fly.io). In the startup script for my Elixir release, I add the following annotations so that every span has information about the server and environment it is executed in. Note that there are more options that you can set, see the [OpenTelemetry semantic conventions for the cloud](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/resource/semantic_conventions/cloud.md)
+
+```bash
+# my_app/rel/overlays/bin/server
+
+#!/bin/sh
+cd -P -- "$(dirname -- "$0")"
+
+read -r -d '' attributes << EOM
+service.name=${FLY_APP_NAME},
+service.instance.id=${FLY_ALLOC_ID},
+cloud.provider=fly_io,
+cloud.region.id=${FLY_REGION}
+EOM
+
+readonly OTEL_RESOURCE_ATTRIBUTES=`echo $attributes | sed 's/[[:space:]]//g'`
+
+OTEL_RESOURCE_ATTRIBUTES=${OTEL_RESOURCE_ATTRIBUTES} \
+    PHX_SERVER=true \
+    exec ./bitfo start
+```
+
+3. Last but not least, we must configure `opentelemtry` to process our spans
+
+In dev and test, we're not going to send our spans anywhere, so they become a noop
+
+```elixir
+# my_app/config/{dev,test}.exs
+
+# Use the noop tracer in dev
+config :opentelemetry,
+       :tracer,
+       :otel_tracer_noop
+```
+
```

Comment:
Question: Is there any other option for verifying things in the development environment? For instance, the Elixir `:telemetry_metrics` package provides a [ConsoleReporter](https://hexdocs.pm/telemetry_metrics/Telemetry.Metrics.ConsoleReporter.html#content) which is useful mostly in the dev environment for debugging metrics configurations and discovering event data.

I'm not that familiar with OpenTelemetry however I imagine that in most cases logging traces to stdout probably doesn't make a lot of sense, but is there anything like that, or anything similar conceptually?

Reply:
I'm almost sure one can setup a docker container as the trace collector do OTel

Reply:
Yes there is definitely a console exporter. I was being lazy here, will add! Good call 💪🏼